nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/firehose/PutKinesisFirehose.java

```diff
@@ -0,0 +1,184 @@
+            Arrays.asList(KINESIS_FIREHOSE_DELIVERY_STREAM_NAME, MAX_BUFFER_INTERVAL,
+                  MAX_BUFFER_SIZE, BATCH_SIZE, REGION, ACCESS_KEY, SECRET_KEY, CREDENTIALS_FILE, AWS_CREDENTIALS_PROVIDER_SERVICE, TIMEOUT,
+                  PROXY_HOST,PROXY_HOST_PORT));
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return properties;
+    }
+
+    @Override
+    protected Collection<ValidationResult> customValidate(final ValidationContext validationContext) {
+        final List<ValidationResult> problems = new ArrayList<>(super.customValidate(validationContext));
+        final boolean batchSizeSet = validationContext.getProperty(BATCH_SIZE).isSet();
+
+        if ( batchSizeSet) {
+           int batchSize = validationContext.getProperty(BATCH_SIZE).asInteger();
+           if ( batchSize < 1 || batchSize > 500 ) {
+                problems.add(new ValidationResult.Builder().input("Batch Size").valid(false).explanation("Batch size must be between 1 and 500 but was " + batchSize).build());
+           }
+        }
+
+        final boolean maxBufferIntervalIsSet = validationContext.getProperty(MAX_BUFFER_INTERVAL).isSet();
+        if ( maxBufferIntervalIsSet) {
+           int maxBufferInterval = validationContext.getProperty(MAX_BUFFER_INTERVAL).asInteger();
+           if ( maxBufferInterval < 60 || maxBufferInterval > 900 ) {
+                problems.add(new ValidationResult.Builder().input("Max Buffer Interval").valid(false)
+                   .explanation("Max Buffer Interval must be between 60 and 900 seconds but was " + maxBufferInterval).build());
+           }
+        }
+
+        final boolean maxBufferSizeIsSet = validationContext.getProperty(MAX_BUFFER_SIZE).isSet();
+        if ( maxBufferSizeIsSet) {
+           int maxBufferSize = validationContext.getProperty(MAX_BUFFER_SIZE).asInteger();
+           if ( maxBufferSize < 1 || maxBufferSize > 128 ) {
+                problems.add(new ValidationResult.Builder().input("Max Buffer Size").valid(false).explanation("Max Buffer Size must be between 1 and 128 (mb) but was " + maxBufferSize).build());
+           }
+        }
+        return problems;
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) {
+        final int batchSize = context.getProperty(BATCH_SIZE).asInteger();
+
+        List<FlowFile> flowFiles = session.get(batchSize);
+        if (flowFiles == null || flowFiles.size() == 0) {
+            return;
+        }
+
+        final String firehoseStreamName = context.getProperty(KINESIS_FIREHOSE_DELIVERY_STREAM_NAME).getValue();
+        final AmazonKinesisFirehoseClient client = getClient();
+
+        try {
+            List<Record> records = new ArrayList<>();
+
+            // Prepare batch of records
+            for (int i = 0; i < flowFiles.size(); i++) {
+                final ByteArrayOutputStream baos = new ByteArrayOutputStream();
```

Comment:
This section is a bit concerning as we are taking the contents of the FlowFiles and putting these items into memory which is going to impact heap.  According to the Kinesis docs (http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/kinesisfirehose/AmazonKinesisFirehoseClient.html) it appears the maximum value this can support is 1,000 kB raw (before encoding occurs).  Given the default batch size, even if constrained, would be ~250MB/instance of exertion on the heap.  

A couple of points:

- Not overly familiar with the Kinesis API but is there a way to kind of stream these items directly from the session so we don't have to load them into memory.
- It would likely be worthwhile to also inspect each flowfile's size to ensure it does not violate the 1000kB supported.  Doing some quick stepping through, it does not appear this is even detected client side.